Sources/PostgresKit/PostgresDataTranslation.swift

```diff
@@ -0,0 +1,365 @@
+                /// Glacial path: Attempt to decode as plain JSON.
+                guard cell.dataType == .json || cell.dataType == .jsonb else {
+                    throw DecodingError.dataCorrupted(.init(codingPath: codingPath, debugDescription: "Unable to interpret value of PSQL type \(cell.dataType): \(cell.bytes.map { "\($0)" } ?? "null")"))
+                }
+                if cell.dataType == .jsonb, cell.format == .binary, let buffer = cell.bytes {
+                    // TODO: Un-hardcode this magic knowledge of the JSONB encoding
+                    return try context.jsonDecoder.decode(T.self, from: buffer.getSlice(at: buffer.readerIndex + 1, length: buffer.readableBytes - 1) ?? .init())
+                } else {
+                    return try context.jsonDecoder.decode(T.self, from: cell.bytes ?? .init())
+                }
+            } catch let error as PostgresDecodingError {
+                /// We effectively transform PostgresDecodingErrors into plain DecodingErrors here, mostly so the full
+                /// coding path, which gives us the original type(s) involved, is preserved.
+                let context = DecodingError.Context(codingPath: codingPath, debugDescription: "\(String(reflecting: error))", underlyingError: error)
+                switch error.code {
+                case .typeMismatch: throw DecodingError.typeMismatch(T.self, context)
+                case .missingData: throw DecodingError.valueNotFound(T.self, context)
+                default: throw DecodingError.dataCorrupted(context)
+                }
+            }
+        }
+    }
+    
+    static func encode<T: Encodable, E: PostgresJSONEncoder>(
+        value: T,
+        in context: PostgresEncodingContext<E>,
+        to bindings: inout PostgresBindings,
+        file: String = #fileID, line: Int = #line
+    ) throws {
+        /// Preferred modern fast-path: Direct conformance to ``PostgresEncodable``
+        if let fastPathValue = value as? any PostgresEncodable {
+            try bindings.append(fastPathValue, context: context)
+        }
+        /// Legacy "fast"-path: Direct conformance to ``PostgresDataConvertible``; use is deprecated.
+        else if let legacyPathValue = value as? any PostgresDataTranslation.PostgresLegacyDataConvertible {
+            guard let legacyData = legacyPathValue.postgresData else {
+                throw EncodingError.invalidValue(value, .init(codingPath: [], debugDescription: "Couldn't get PSQL encoding from value '\(value)'"))
+            }
+            bindings.append(legacyData)
+        }
+        /// Slow path: Descend through the ``Encodable`` machinery until we fail or find something we can convert.
+        else {
+            try bindings.append(self.encode(codingPath: [], userInfo: [:], value: value, in: context, file: file, line: line))
+        }
+    }
+    
+    internal /*fileprivate*/ static func encode<T: Encodable, E: PostgresJSONEncoder>(
```

Comment:
Can we remove the file private from here?

Reply:
This comment follows the same style used by NIO, used to indicate "this _should_ be `[file]private` but is `internal` so unit tests can access it.